New page for offenses and slashes (#5878)
w3f/polkadot-wiki: docs/learn/learn-offenses.md

keyword: [nominate, nominator, offenses, slashes, validator, equivocation, disabling]
slug: ../learn-offenses
---

:::info Content subject to change

The material provided here is based on the changes introduced by Step 2 of the _Disabling_ feature.
See [this page](https://github.com/orgs/paritytech/projects/119/views/15?pane=issue&itemId=61684472)
for more information.

:::

{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is a public permissionless network.
As such, it has a mechanism to disincentivize offenses and incentivize good behavior. Below, you can
find a summary of punishments for specific offenses:

|               Offense                | [Slash (%)](#slashing) | [On-chain Disabling](#disabling) | Off-chain Disabling | [Rep](#rep) |
| :----------------------------------: | :--------------------: | :------------------------------: | :-----------------: | :---------: |
|           Backing Invalid            |          100%          |               Yes                | Yes (High Priority) |     No      |
|           ForInvalid Vote            |           -            |                No                | Yes (Mid Priority)  |     No      |

> |         AgainstInvalid Vote          |           -            |                No                | Yes (Low Priority)  |     No      |

Reviewer: The https://github.com/paritytech/polkadot-sdk/issues/4359 has a AgainstValid vote here. 
```suggestion
|         AgainstValid Vote          |           -            |                No                | Yes (Low Priority)  |     No      |
```

This is also called a "Spam Dispute". It means that we had a perfectly valid block but someone (troll or an attacker) said it was invalid and started an escalation procedure (dispute) and everyone has to check it now to verify it. All other nodes in the validator set would probably say the block is valid and that sole attacker would loose the dispute and be charged with the AgainstValid Vote offence. Currently there is no slash associated with it but they will most likely be enabled in the future. Same for ForInvalid.

| GRANDPA / BABE / BEEFY Equivocations |       0.01-100%        |               Yes                |         No          |     No      |
|    Seconded + Valid Equivocation     |           -            |                No                |         No          |     No      |
|     Double Seconded Equivocation     |           -            |                No                |         No          |     Yes     |

## Offenses

:::info Learn more about the parachain protocol

To better understand the terminology used for offenses, it is recommended to get familiar with the
[parachain protocol](./learn-parachains-protocol.md).

:::

> On {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}, there are six main offenses as

Reviewer: ```suggestion
On {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}, there are six main validator offenses as
```

Trying to make the distinction so people don't think that a random slash on a smart contract on moonbeam is the same as this protocol level validator slash.

shown below.

- Backing Invalid: A para-validator is backing an invalid block.
- ForInvalid Vote: A validator (secondary checker) votes in favor of an invalid block.

> - AgainstInvalid Vote: A validator (secondary checker) is voting against an invalid block.

Reviewer: Update based on the above AgainstValid comment. 

Maybe 
```suggestion
- AgainstValid Vote: A validator (secondary checker) is voting against a valid block and wasting network resources.
```

Feel free to expand it with how much context you believe is useful.

- Equivocation: A validator produces two or more of the same block or vote.
  - GRANDPA and BEEFY Equivocation: A validator signs two or more votes in the same round on
    different chains.
  - BABE Equivocation: A validator produces two or more blocks on the Relay Chain in the same time
    slot.
- Seconded + Valid Equivocation: **TODO**

> - Double Seconded Equivocation: **TODO**

Reviewer: I'm considering remove those two from the wiki here. Yes they are technically offences but they are quite in-depth and we don't really slash for them. We very much could slash because honest nodes should never do it, but not slashing doesn't compromise security. 

Let me explain and then we can decide if we want it here or not. Double Seconded Equivocation is a bit easier to understand.

**Double Seconded Equivocation:** 

https://paritytech.github.io/polkadot-sdk/book/node/backing/candidate-backing.html#candidate-backing:
> The Candidate Backing subsystem ensures every parablock considered for relay block inclusion has been seconded by at least one validator, and approved by a quorum.

That means that within a backing group of 5 para-validators we can get at most 5 backed parablocks. Each parablock requires exactly 1 seconded and at least 2 more valid votes from the 5 potential backers. This makes an upper bound on the number of parablocks the system has to deal with while still allowing some choice for relay chain block authors. Essentially backers need to decide which parablock to second and they can't second another. If another seconding vote would be found we will punish them (somewhat lightly as of now but there's little to gain from this).

All of this is made slightly more complicated with async backing as it is not longer 1 candidate per relay chain block as backers can back blocks "into the future" optimistically. You can send people here for the advanced async backing perspective: https://paritytech.github.io/polkadot-sdk/book/node/backing/statement-distribution.html#seconding-limit

**Seconded + Valid Equivocation:**

This is a case when the malicious node first seconds something (takes absolute responsibility for it) but then only pretends he was someone that just said it's correct after someone else took responsibility. That is a straight up lie (equivocation). A node could use that tactic to potentially escape from responsibility but once we notice the two conflicting votes the offence is reported.

Author: this is quite important information IMO. I will add it and maybe we can decide later on to remove it ;)

### Equivocation

Equivocation events can occur when a validator produces two or more of the same block; under this
condition, it is referred to as a BABE equivocation. Equivocation may also happen when a validator
signs two or more of the same consensus vote; under this condition, it is referred to as a GRANDPA
Equivocation. Equivocations usually occur when duplicate signing keys reside on the validator host.
If keys are never duplicated, the probability of an equivocation slash decreases to near 0.

> 

Reviewer: ```suggestion
### Equivocation (Conflicting Statements)

Equivocation occurs when a validator produces statements that conflict with each other.

For instance as a block author appointed by BABE only a single block should be authored for the given slot and if two or more are authored they are in conflict with each other. This would be a BABE Equivocation Offence.

In BEEFY & GRANDPA validators are expected to cast a single vote for the block they believe is the best, but if they are found with 2 or more votes for different blocks in means they tried to confuse the network with conflicting statements and when found out this will be a BEEFY/GRANDPA Equivocation Offence.

Equivocations usually occur when duplicate signing keys reside on the validator host.
If keys are never duplicated, the probability of an honest equivocation slash decreases to near 0.
```

Feel free to format above differently

## Punishments

On {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}, offenses to the network can be
punished depending on their severity. There are three main punishments: slashing, disabling, and

> rep.

Reviewer: ```suggestion
reputation changes.
```

### Slashing

**Slashing** removes part of a validator’s total stake (own + nominated) and can range from as

> little as 0.01% or rise to 100%. In all instances, slashes are accompanied by a loss of nominators.

Reviewer: >  In all instances, slashes are accompanied by a loss of nominators.

Confusing wording. What do you mean by that? Nominators are not auto unsubscribing when the validator is slashed. Nominators get slashed when validator is slashed. We can have a solo validator with no nominators (only own stake). Unsure what it aimed to convey

Reviewer 2: This was valid till a year ago https://forum.polkadot.network/t/kusama-era-4543-slashing/1410/2?u=radha and was updated after a slashing event on Kusama in Dec 2022. We will update the Wiki to reflect this.

Author: yep no automatic loss of nominators, corrected and unified with the following paragraph essentially repeating the same stuff

Slashing only occurs for active validations for a given nominator, and slashes are not mitigated by
having other inactive or waiting nominations. They are also not mitigated by the validator operator
running separate nodes; each node is considered its own entity for slashing purposes.

:::info Multiple Active Nominations

In rare instances, with very large bonds, a nominator may actively nominate several validators in a
single era. In this case, the slash is proportionate to the amount staked to that specific
validator. Note that you cannot control the percentage of stake allocated to each validator or
choose who your active validator will be (except in the trivial case of nominating a single
validator). Staking allocations are controlled by the [Phragmén algorithm](learn-phragmen.md).

:::

Once a validator gets slashed, it goes into the state as an "unapplied slash". You can check this
via
[Polkadot-JS UI](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Frpc.polkadot.io#/staking/slashes).
The UI shows it per validator, followed by all the affected nominators and the amounts. While
unapplied, a governance proposal can be made to reverse it during this period

> ({{ polkadot: <RPC network="polkadot" path="consts.staking.bondingDuration" defaultValue={28} filter="erasToDays"/> :polkadot }}{{ kusama: <RPC network="kusama" path="consts.staking.bondingDuration" defaultValue={28} filter="erasToDays"/> :kusama }}

Reviewer: it is 1 day lower than bonding duration so 27

Author: changed

days). After the grace period, the slashes are applied.

A slash may occur under the circumstances below:

1.  Equivocation – A slash of 0.01% is applied with as little as a single evocation. The slashed
    amount increases to 100% incrementally as more validators also equivocate.
2.  Malicious action – This may result from a validator trying to represent the contents of a block

>     falsely . Slashing penalties of 100% may apply.

Reviewer: Let's make it coherent with the table above. First of all Equivocations can be malicious actions so that distinction made no sense anyway. Let's keep Equivocations above and change Malicious Action.

Maybe "Dispute Slashes" would be appropriate. This would cover the 3 types of offences from the table (Backing Invalid, ForInvalid and AgainstValid votes). Rn we dont slash for ForInvalid and AgainstValid votes but that might happen soon-ish (year or so).

Author: done as suggested

3.  Application related (bug or otherwise) – The amount is unknown and may manifest as scenarios 1

>     and 2 above.

Reviewer: Not even sure what this is. Validator/staking slashes are only caused by Equivocation or Backing Invalid Parablocks. I'd remove that 3rd option here.

It might be referring to external non staking/validator slashes. But we should mention them separately.

1.  Cloning a server, i.e., copying all contents when migrating to new hardware. This action should
    be avoided. If an image is desired, it should be taken before keys are generated.
2.  High Availability (HA) Systems – Equivocation can occur if there are any concurrent operations,
    either when a failed server restarts or if a false positive event results in both servers being
    online simultaneously. HA systems are to be treated with extreme caution and are not advised.
3.  The keystore folder is copied when attempting to copy a database from one instance to another.  
    It is important to note that equivocation slashes occur with a single incident. This can happen
    if duplicated keystores are used for only a few seconds. A slash can result in losing nominators
    and funds, removal from the Thousand Validator Programme, and reputational damage.

See the next section to understand how slash amounts for equivocations are calculated. If you want
to know more details about slashing, please look at our
[research page](https://research.web3.foundation/Polkadot/security/slashing/amounts).

#### Slash Calculation for Equivocation

> Both GRANDPA and BABE equivocation use the same formula for calculating the slashing penalty:

Reviewer: ```suggestion
GRANDPA, BABE and BEEFY equivocation use the same formula for calculating the slashing penalty:
```

#### Slashing Across Eras

There are three main difficulties to account for with slashing in NPoS:

- A nominator can nominate multiple validators and be slashed via any of them.
- Until slashed, the stake is reused from era to era. Nominating with N coins for E eras in a row
  does not mean you have N\*E coins to be slashed - you've only ever had N.
- Slashable offenses can be found after the fact and out of order.

To balance this, we only slash for the maximum slash a participant can receive in some time period
rather than the sum. This ensures protection from overslashing. Likewise, the period over which
maximum slashes are computed is finite, and the validator is chilled with nominations withdrawn
after a slashing event, as stated in the previous section. This prevents rage-quit attacks in which,
once caught misbehaving, a participant deliberately misbehaves more because their slashing amount is

> already maxed out.

Reviewer: Very stale. Withdrawing nominations has not been there for a long time. Recently we also made it so validators are not involuntarily chilled.

```suggestion
To balance this, we only slash for the maximum slash a participant can receive in some time period
rather than the sum. This ensures protection from overslashing.
```

### Disabling

> **Disabling** stops validators from performing specific actions. Disabling is further divided into:

Reviewer: ```suggestion
**Disabling** stops validators from performing specific actions after they committed an offence. Disabling is further divided into:
```

> - On-chain disabling lasts for a whole era and stops validators from block authoring, backing
>   through runtime filtering, initiating a dispute, and backing. It also makes other nodes ignore
>   backing statements.

Reviewer: ```suggestion
- On-chain disabling lasts for a whole era and stops validators from block authoring, backing and initiating a dispute.
```

deduplication + simplified. Backing statement filtering is there but it is an optimization that is not worth mentioning.

- Off-chain disabling lasts for a session, is caused by losing a dispute, and stops validators from
  initiating a dispute.

Off-chain disabling is always a lower priority than on-chain disabling. Off-chain disabling

> prioritizes disabling first backers, then ForInvalid, then AgainstValid.

Reviewer: ```suggestion
prioritizes disabling first backers then approval checkers.
```
Less detail might be more understandable here.

> ### Rep

Reviewer: ```suggestion
### Reputation Changes
```

> **TODO**

Reviewer: Some minor offences often connected to spamming are only punished by Networking Reputation Changes. When validators connect to each other they use a reputation metric for each of their peers. If our peers provide valuable data and behave properly we add reputation and if they provide us with faulty or spam data we reduce their reputation. If a validator looses enough reputation the peer will temporarily close the channel. This helps in fighting against DoS (Denial of Service) attacks. The consequences of closing the channel might be varied but in general performing validator tasks will be harder so it might result in lower validators rewards.

Author: added

- **AgainstValid Vote:** A validator (secondary checker) is voting against a valid block and wasting
  network resources.
- **Equivocation:** A validator produces two or more of the same block or vote.
  - GRANDPA and BEEFY Equivocation: A validator signs two or more votes in the same round on
    different chains.
  - BABE Equivocation: A validator produces two or more blocks on the Relay Chain in the same time
    slot.
- **Double Seconded Equivocation:** Within a backing group of 5 para-validators we can get at most 5
  backed parablocks. Each parablock requires exactly 1 seconded and at least 2 more valid votes from
  the 5 potential backers. This makes an upper bound on the number of parablocks the system has to
  deal with while still allowing some choice for relay chain block authors. Backers must decide
  which parablock to second and they cannot second another. If another seconding vote would be found
  we will punish them (somewhat lightly as of now but there's little to gain from this). All of this
  is made slightly more complicated with [asynchronous backing](./learn-async-backing.md) as it is

>   not longer 1 candidate per relay chain block as backers can back blocks "into the future"

Reviewer: ```suggestion
  no longer 1 candidate per relay chain block as backers can back blocks "into the future"
```
